R/integrate_sces.R

```diff
@@ -0,0 +1,180 @@
+                           seed = NULL,
+                           ...) {
+
+  # Set seed
+  set.seed(seed)
+
+  # make sure that input is a SingleCellExperiment
+  if(!is(merged_sce, "SingleCellExperiment")){
+    stop("The `merged_sce` must be a SingleCellExperiment object created with `scpcaTools::merge_sce_list().`")
+  }
+
+
+  # Check integration_method
+  integration_method <- match.arg(integration_method)
+
+  # Check batch_column
+  if (!(batch_column %in% names(colData(merged_sce)))) {
+    stop("The provided `batch_column` column must be in the `merged_sce` colData.")
+  }
+  # Ensure there are no NAs
+  batches <- colData(merged_sce)[[batch_column]]
+  if( sum(is.na(batches)) > 0 ) {
+    stop("`NA` values are not allowed in the `batch_column`.")
+  }
+  # Ensure >=2 batches
+  if( length(unique(batches)) < 2 ) {
+    stop("At least two batches are required for integration.")
+  }
+
+
+  # Perform integration with specified method
+  if (integration_method == "fastMNN") {
+
+    if (length(covariate_cols) >= 1) {
+      warning("fastMNN cannot use additional covariates, so `covariate_cols` will be ignored.")
+    }
+
+    integrated_sce <- integrate_fastmnn(merged_sce,
+                                        batch_column,
+                                        ...)
+    # Add corrected expression to merged_sce if specified
+    if (return_corrected_expression) {
+      assay(merged_sce, "fastMNN_corrected") <- assay(integrated_sce, "reconstructed")
+    }
+    # define PCs
+    integrated_pcs <- reducedDim(integrated_sce, "corrected")
+  }
+  if (integration_method == "harmony") {
+
+    # here the result is the PCs:
+    integrated_pcs <- integrate_harmony(merged_sce,
+                                        batch_column,
+                                        covariate_cols,
+                                        ...)
+  }
+
+  # Add PCs from integration into merged_sce (name is lowercase)
+  reducedDim(merged_sce, glue::glue("{integration_method}_PCA")) <- integrated_pcs
+
+  # Return the merged_sce with integration results included
+  return(merged_sce)
+}
+
+
+
+#' Integrate a list of SingleCellExperiment (SCE) objects using the fastMNN
+#' approach from the `batchelor` package.
+#' Source: http://www.bioconductor.org/packages/release/bioc/html/batchelor.html
+#'
+#' @param merged_sce A merged SCE object as prepared by `scpcaTools::merge_sce_list()`.
+#' @param batch_column The variable in `merged_sce` indicating batches.
+#' @param ... Additional arguments to pass into `batchelor::fastMNN()`
+#'
+#' @return An integrated SCE object produced by `fastMNN`
+#'
+#' @import SingleCellExperiment
+integrate_fastmnn <- function(merged_sce,
+                              batch_column,
+                              ...) {
+
+  # Check the merged_sce for logcounts
+  if (!("logcounts" %in% names(assays(merged_sce)))) {
+    stop("The `merged_sce` object requires a `logcounts` assay for fastMNN integration.")
+  }
```

Comment:
We need to check that batchelor is installed, since it is not a requirement for scpcaTools in general. Similarly, we will need a check for `harmony`
```suggestion
  # Check that batchelor is installed
  if (!requireNamespace("batchelor", quietly = TRUE)) { 
    stop("The batchelor package must be installed to use fastMNN.") 
  }

  # Check the merged_sce for logcounts
  if (!("logcounts" %in% names(assays(merged_sce)))) {
    stop("The `merged_sce` object requires a `logcounts` assay for fastMNN integration.")
  }
```